cmd/promtool/rules.go

```diff
@@ -0,0 +1,197 @@
+	return &ruleImporter{
+		logger:      logger,
+		config:      config,
+		apiClient:   apiClient,
+		appender:    appender,
+		ruleManager: rules.NewManager(&rules.ManagerOptions{}),
+	}
+}
+
+// loadGroups parses groups from a list of recording rule files.
+func (importer *ruleImporter) loadGroups(ctx context.Context, filenames []string) (errs []error) {
+	groups, errs := importer.ruleManager.LoadGroups(importer.config.EvalInterval, labels.Labels{}, filenames...)
+	if errs != nil {
+		return errs
+	}
+	importer.groups = groups
+	return nil
+}
+
+// importAll evaluates all the recording rules and creates new time series and writes them to disk in blocks.
+func (importer *ruleImporter) importAll(ctx context.Context) (errs []error) {
+	for name, group := range importer.groups {
+		level.Info(importer.logger).Log("backfiller", fmt.Sprintf("processing group, name: %s", name))
+
+		stimeWithAlignment := group.EvalTimestamp(importer.config.Start.UnixNano())
+		for i, r := range group.Rules() {
+			level.Info(importer.logger).Log("backfiller", fmt.Sprintf("processing rule %d, name: %s", i+1, r.Name()))
+			if err := importer.importRule(ctx, r.Query().String(), r.Name(), r.Labels(), stimeWithAlignment, importer.config.End, group.Interval()); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	if err := importer.appender.flushAndCommit(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+// importRule queries a prometheus API to evaluate rules at times in the past.
+func (importer *ruleImporter) importRule(ctx context.Context, ruleExpr, ruleName string, ruleLabels labels.Labels, start, end time.Time, interval time.Duration) error {
+
+	// This loop breaks up the calls to the QueryRange API into 2 hr chunks so that we
+	// don't ever request too much data or take to long to process to avoid timeout.
+	for start.Before(end) {
+		currentBlockEnd := start.Add(time.Duration(tsdb.DefaultBlockDuration) * time.Millisecond)
+		if currentBlockEnd.After(end) {
+			currentBlockEnd = end
+		}
+
+		val, warnings, err := importer.apiClient.QueryRange(ctx,
+			ruleExpr,
+			v1.Range{
+				Start: start,
+				End:   end,
+				Step:  interval,
+			},
+		)
+		if err != nil {
+			return err
+		}
+		if warnings != nil {
+			level.Warn(importer.logger).Log("backfiller", fmt.Sprintf("warnings QueryRange api: %v", warnings))
+		}
+
+		var matrix model.Matrix
+		switch val.Type() {
+		case model.ValMatrix:
+			matrix = val.(model.Matrix)
+			for _, sample := range matrix {
+				currentLabels := make(labels.Labels, 0, len(sample.Metric)+len(ruleLabels)+1)
+				currentLabels = append(currentLabels, labels.Label{
+					Name:  labels.MetricName,
+					Value: ruleName,
+				})
+
+				currentLabels = append(currentLabels, ruleLabels...)
+
+				for name, value := range sample.Metric {
+					currentLabels = append(currentLabels, labels.Label{
+						Name:  string(name),
+						Value: string(value),
+					})
+				}
+				for _, value := range sample.Values {
```

Comment:
It just occurred to me that we need to insert stale markers here when there's a sample missing for a given evaluation time, but the previous evaluation time had a sample.

Reply:
oh interesting. Let me look into adding that.